File: API/apps/routes/vouchers.py
from flask import Blueprint
from apps.services.Vouchers import Voucher
from apps.services.BaseServices import token_auth

voucher = Blueprint("voucher", __name__, url_prefix="/api/voucher")
voucher.before_request(lambda: token_auth.login_required(lambda: None)())

@voucher.route("", methods=["GET"])
def _gettVoucher(): return Voucher().getVouchers()

@voucher.route("/<int:tipe>", methods=["POST"])
def _insertVoucher(tipe): return Voucher().insertVoucher(tipe)

@voucher.route("/<int:tipe>", methods=["PUT"])
def _updateVoucher(tipe): return Voucher().updateVoucher(tipe)

@voucher.route("/<int:tipe>", methods=["DELETE"])
def _deleteVoucher(tipe): return Voucher().deleteVoucher(tipe)

@voucher.route("/tolak-voucher", methods=["POST"])
def _decline_voucher(): return Voucher().declineProductVoucher()

@voucher.route("/tolak-voucher-regular", methods=["POST"])
def _decline_voucher_regular(): return Voucher().declineRegularVoucher()

@voucher.route("/use-voucher", methods=["POST"])
def _use_voucher(): return Voucher().useVoucher()

@voucher.route("/get-v1-regular")
def _1(): return Voucher().getVoucher_1_reguler()

@voucher.route("/get-v2-product/<int:id_produk>")
def _2(id_produk): return Voucher().getVoucher_2_produk(id_produk)

@voucher.route("/get-v2-regular")
def _2r(): return Voucher().getVoucher_2_reguler()

@voucher.route("/get-v3-product/<int:id_produk>")
def _3(id_produk): return Voucher().getVoucher_3_produk(id_produk)

@voucher.route("/get-v3-regular")
def _3r(): return Voucher().getVoucher_3_reguler()

@voucher.route("/get-voucher-by-id/<int:tipe>")
def _getVoucherById(tipe): return Voucher().getVoucherById(tipe)

@voucher.route("/get-v2-product-all")
def _get_v2_product_all(): return Voucher().getVoucher_2_produk_all()

@voucher.route("/get-v3-product-all")
def _get_v3_product_all(): return Voucher().getVoucher_3_produk_all()

File: 9budimas-sales-canvas/src/services/vouchers.js
import { fetchWithAuth, buildUrlWithParams, handleError, localDisk } from "../lib/utils";
import { baseService } from "./baseService";

class VoucherService extends baseService {
  #baseGetUrl;

  constructor() {
    super();
    this.#baseGetUrl = `/api/voucher`;
    this.serviceName = "voucher";
    this.endpoints = this.#baseGetUrl;
  }

  _getCurrentUser() {
    return localDisk.getLocalStorage("user");
  }

  async _fetch(method, url, payload = null, { wrapInArray = false, errorMessage } = {}) {
    try {
      const res = await fetchWithAuth(method, url, payload);
      return wrapInArray ? (Array.isArray(res) ? res : [res]) : res;
    } catch (error) {
      handleError(error, errorMessage);
    }
  }

  async getVouchers() {
    return this._fetch("GET", this.#baseGetUrl, null, { errorMessage: "Gagal mengambil data voucher" });
  }

  async insertVoucher(tipe, data) {
    return this._fetch("POST", `${this.#baseGetUrl}/${tipe}`, data, { errorMessage: "Gagal menambahkan voucher" });
  }

  async updateVoucher(tipe, data) {
    return this._fetch("PUT", `${this.#baseGetUrl}/${tipe}`, data, { errorMessage: "Gagal mengupdate voucher" });
  }

  async deleteVoucher(tipe) {
    return this._fetch("DELETE", `${this.#baseGetUrl}/${tipe}`, null, { errorMessage: "Gagal menghapus voucher" });
  }

  async declineProductVoucher(data) {
    return this._fetch("POST", `${this.#baseGetUrl}/tolak-voucher`, data, { errorMessage: "Gagal menolak voucher produk" });
  }

  async declineRegularVoucher(data) {
    return this._fetch("POST", `${this.#baseGetUrl}/tolak-voucher-regular`, data, { errorMessage: "Gagal menolak voucher regular" });
  }

  async useVoucher(data) {
    return this._fetch("POST", `${this.#baseGetUrl}/use-voucher`, data, { errorMessage: "Gagal menggunakan voucher" });
  }

  async getVoucherV1Regular() {
    return this._fetch("GET", `${this.#baseGetUrl}/get-v1-regular`, null, { errorMessage: "Gagal mengambil voucher reguler V1" });
  }

  async getVoucherV2Product(id_produk, params = {}) {
    const url = buildUrlWithParams(`${this.#baseGetUrl}/get-v2-product/${id_produk}`, params);
    return this._fetch("GET", url, null, { errorMessage: "Gagal mengambil voucher produk V2" });
  }

  async getVoucherV2Regular() {
    return this._fetch("GET", `${this.#baseGetUrl}/get-v2-regular`, null, { errorMessage: "Gagal mengambil voucher reguler V2" });
  }

  async getVoucherV3Product(id_produk, params = {}) {
    const url = buildUrlWithParams(`${this.#baseGetUrl}/get-v3-product/${id_produk}`, params);
    return this._fetch("GET", url, null, { errorMessage: "Gagal mengambil voucher produk V3" });
  }

  async getVoucherV3Regular() {
    return this._fetch("GET", `${this.#baseGetUrl}/get-v3-regular`, null, { errorMessage: "Gagal mengambil voucher reguler V3" });
  }

  async getVoucherById(tipe) {
    return this._fetch("GET", `${this.#baseGetUrl}/get-voucher-by-id/${tipe}`, null, { errorMessage: "Gagal mengambil voucher by ID" });
  }

  async getVoucherV2ProductAll() {
    return this._fetch("GET", `${this.#baseGetUrl}/get-v2-product-all`, null, { errorMessage: "Gagal mengambil semua voucher produk V2" });
  }

  async getVoucherV3ProductAll() {
    return this._fetch("GET", `${this.#baseGetUrl}/get-v3-product-all`, null, { errorMessage: "Gagal mengambil semua voucher produk V3" });
  }
}

const voucherService = new VoucherService();
export { voucherService };
